table/block.cc

```diff
@@ -505,9 +513,50 @@ bool IndexBlockIter::PrefixSeek(const Slice& target, uint32_t* index) {
   }
 }
 
+// NOTE: in suffix seek, if the key is not found in the restart intervals,
+// the iterator will simply be set as "invalid", rather than returning
+// the key that is just pass the target key.
+// return value: found
+bool DataBlockIter::HashSeek(const Slice& target) {
+  assert(data_block_hash_index_);
+  Slice user_key = ExtractUserKey(target);
+  std::unique_ptr<DataBlockHashIndexIterator> data_block_hash_iter(
+      data_block_hash_index_->NewIterator(user_key));
```

Comment:
We need to get rid of this malloc here.

Reply:
I may have an iterator instance embedded in the `DataBlockIter` to avoid the extra malloc, and initialize it every time I use it. Would this work?

The trade-off is, a new `DataBlockIter` is malloc-ed every time `BlockBasedTable::Get()` is called. If I can embed the `DataBlockHashIndexIterator` into `DataBlockIter` we save half of the malloc but lose a little space.

Reply:
Please don't do that. 
Instead, you can allocate it in the stack. Something like what we do in DataBlockIterator
```
DataBlockHashIndexIterator iter;
data_block_hash_index_->NewIterator(user_key, &iter);
```

Reply:
Another way is totally bypass BlockIter interface, and have a new call path from `BlockBasedTable::Get()`, to `Block::Get()` (TODO). I think this is doable, and it also solves the extra memory access from the restart index to restart offset as described below.